### ISSUE TITLE
12

### ISSUE THREAD

**Issue by dingding800** [dingding800/HomeAutomation#15]
import pandas as pd
import numpy as np
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import StratifiedKFold
from sklearn.metrics import roc_auc_score

# Step 1: Simulate data (each variable scored 1–4, bearerShare is binary)
np.random.seed(42)
n_samples = 200
df = pd.DataFrame({
    'ClientType': np.random.randint(1, 5, n_samples),
    'PEP': np.random.randint(1, 5, n_samples),
    'RAI': np.random.randint(1, 5, n_samples),
    'Sanction': np.random.randint(1, 5, n_samples),
    'country': np.random.randint(1, 5, n_samples),
    'sector': np.random.randint(1, 5, n_samples),
    'bearerShare': np.random.randint(0, 2, n_samples),
    'target': np.random.randint(0, 2, n_samples)
})

# Step 2: Logistic regression with 5-fold cross-validation
X = df[['ClientType', 'PEP', 'RAI', 'Sanction', 'country', 'sector', 'bearerShare']]
y = df['target']
auc_scores = []

kfold = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)
for train_idx, val_idx in kfold.split(X, y):
    model = LogisticRegression()
    model.fit(X.iloc[train_idx], y.iloc[train_idx])
    y_pred = model.predict_proba(X.iloc[val_idx])[:, 1]
    auc_scores.append(roc_auc_score(y.iloc[val_idx], y_pred))

average_auc = np.mean(auc_scores)
print(f"\n✅ Average AUC across 5-folds: {average_auc:.4f}")

# Step 3: Information Value (IV) calculation
def calculate_iv(df, feature, target):
    grouped = df.groupby(feature).agg({target: ['sum', 'count']})
    grouped.columns = ['bad', 'total']
    grouped['good'] = grouped['total'] - grouped['bad']
    grouped['dist_bad'] = grouped['bad'] / grouped['bad'].sum()
    grouped['dist_good'] = grouped['good'] / grouped['good'].sum()
    grouped['woe'] = np.log(grouped['dist_bad'] / grouped['dist_good']).replace([np.inf, -np.inf], 0)
    grouped['iv'] = (grouped['dist_bad'] - grouped['dist_good']) * grouped['woe']
    return grouped['iv'].sum()

print("\n📊 IV (Information Value) per variable:")
for col in X.columns:
    iv = calculate_iv(df, col, 'target')
    print(f"{col:15}: IV = {iv:.4f}")

# Step 4: PSI (Population Stability Index)
# Simulate a new population to compare against (e.g., next month's data)
df_new = df.copy()
for col in X.columns:
    if col == 'bearerShare':
        df_new[col] = np.random.randint(0, 2, n_samples)
    else:
        df_new[col] = np.random.randint(1, 5, n_samples)

def calculate_psi(expected, actual):
    psi_total = 0
    values = sorted(set(expected) | set(actual))
    for v in values:
        e_pct = (expected == v).sum() / len(expected)
        a_pct = (actual == v).sum() / len(actual)
        e_pct = max(e_pct, 0.0001)
        a_pct = max(a_pct, 0.0001)
        psi = (a_pct - e_pct) * np.log(a_pct / e_pct)
        psi_total += psi
    return psi_total

print("\n📈 PSI (Population Stability Index) per variable:")
for col in X.columns:
    psi = calculate_psi(df[col], df_new[col])
    print(f"{col:15}: PSI = {psi:.4f}")
